src/uct/ib/base/ib_iface.c

```diff
@@ -673,43 +705,84 @@ static ucs_status_t uct_ib_iface_init_pkey(uct_ib_iface_t *iface,
         }
 
         pkey = ntohs(port_pkey);
-        if (!(pkey & UCT_IB_PKEY_MEMBERSHIP_MASK)) {
-            /* if pkey = 0x0, just skip it w/o debug trace, because 0x0
-             * means that there is no real pkey configured at this index */
-            if (pkey) {
-                ucs_trace("skipping send-only pkey[%d]=0x%x on "UCT_IB_IFACE_FMT,
-                          pkey_index, pkey, UCT_IB_IFACE_ARG(iface));
-            }
+        /* if pkey = 0x0, just skip it w/o debug trace, because 0x0
+         * means that there is no real pkey configured at this index */
+        if (pkey == 0) {
+            continue;
+        }
+
+        if ((config->pkey == UCS_HEXUNITS_AUTO) &&
+            !(pkey & UCT_IB_PKEY_MEMBERSHIP_MASK)) {
+            ucs_trace("skipping send-only pkey[%d]=0x%x on "UCT_IB_IFACE_FMT,
+                      pkey_index, pkey, UCT_IB_IFACE_ARG(iface));
+            has_pkey_limited = 1;
             continue;
         }
 
         /* take only the lower 15 bits for the comparison */
-        if ((config->pkey_value == UCS_HEXUNITS_AUTO) ||
-            ((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey_value)) {
+        if ((config->pkey == UCS_HEXUNITS_AUTO) ||
+            ((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey)) {
             iface->pkey_index = pkey_index;
-            iface->pkey_value = pkey;
+            iface->pkey       = pkey;
             pkey_found        = 1;
             break;
         }
     }
 
     if (!pkey_found) {
-        if (config->pkey_value == UCS_HEXUNITS_AUTO) {
-            ucs_error("There is no valid pkey with full membership on "
-                      UCT_IB_IFACE_FMT, UCT_IB_IFACE_ARG(iface));
+        if (config->pkey == UCS_HEXUNITS_AUTO) {
+            if (!allow_limited && has_pkey_limited) {
+                ucs_debug("there is no valid pkey with full membership on "
+                          UCT_IB_IFACE_FMT, UCT_IB_IFACE_ARG(iface));
+                return UCS_INPROGRESS;
```

Comment:
should return UCS_ERR_NO_ELEM if not found

Reply:
yes, but the error code from this function should reflect that this function did not find a pkey
the called may choose to continue if status==UCS_ERR_NO_ELEM, the further action based on the return status is outside the scope of this function

Reply:
yes, makes sense to me